Develop

Anurag19101996 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Added R script to generate DUCKDB for SAU model

## Motivation and Context (link issue)
#1  
#2 

## How Has This Been Tested?
Tests done on local machine to test creation of tables and validate data types for the columns.

Anurag19101996 committed:
Create README.md

Anurag19101996 committed:
Python notebook to generate duckdb

Anurag19101996 committed:
Merge pull request #1 from Anurag19101996/develop-SAU_Duckdb

Add Python notebook to create the DUCKDB for SAU

Anurag19101996 committed:
Delete DUCKDB directory

Anurag19101996 committed:
Python notebook to create DUCKDB file for SAU*

Anurag19101996 committed:
Delete SAU_model.ipynb

Anurag19101996 committed:
Create ReadME.md

Anurag19101996 committed:
R script to generate DuckDB file w/o metadata.

Anurag19101996 committed:
Update ReadME.md

Anurag19101996 committed:
Update ReadME.md

Anurag19101996 committed:
Duckdb File creation R script (with metadata)

Note: The metadata is not based on EML standards

Anurag19101996 committed:
Delete SAU Model/SAU_DUCKDB_Creation_script.R

Anurag19101996 committed:
Update ReadME.md

Anurag19101996 committed:
Update ReadME.md

Anurag19101996 committed:
Update ReadME.md

Anurag19101996 committed:
Merge branch 'Seafood-Globalization-Lab:develop' into develop

Anurag19101996 commented:
@theamarks please review it for SAU model.

#1 #2

theamarks commented on SAU Model/ReadME.md:
```diff
@@ -0,0 +1,28 @@
+Column specifications are sourced from the [ARTIS Database Tables](https://github.com/Seafood-Globalization-Lab/artis-model/wiki/ARTIS-Database-Tables). Metadata is based on these specifications. Discrepancies between table descriptions and source files are noted.
+
+## Notable Discrepancies
+- **Table: `consumption`**
```
I updated all table documentation in the wiki pages based on your findings. I also added a column listing the SQL data type as well. Thanks for bringing this to our attention.

theamarks commented on SAU Model/ReadME.md:
```diff
@@ -0,0 +1,28 @@
+  - **Source File Name**: `consumption_t_capped`
+
+## Data Type Considerations
+- The `hs6` code is described as an integer in the GitHub table specifications but is stored as a string in the DuckDB to preserve leading zeros.
```
I updated the wiki table documentation to address this issue. I am still unsure if we directly `CREATE TABLE` in SQL and `INSERT` directly from a .csv if this data will be preserved with the leading zeros we want. https://github.com/Seafood-Globalization-Lab/artis-model/wiki/ARTIS-Database-Tables#trade

We need to make sure we are taking care of this leading zeros issue in the appropriate location in the pipeline. Is it the table write out step OR reading the .csv into R?

Anurag19101996 replied:
Hi @theamarks, we are loading the data from the table and checking the number of characters in hs6, if its 5 then we add a leading zero.  Its the table write out step.

Anurag19101996 replied:
Hi @theamarks ,

I've updated the R script (SAU_DuckDB_Script_w_metadata_v2.R) to first create blank SQL tables with explicitly defined data types, and then ingest data from the CSVs.

I've added two layers of data type validation to ensure schema consistency:

The SQL tables are created with explicitly defined column names and data types.

Each column in the CSV is coerced to the correct type before insertion.

If there's any mismatch between the defined SQL schema and the actual column types in the data, DuckDB will throw an error — which helps catch issues early.

theamarks commented on SAU Model/SAU_DuckDB_Script_w_metadata.R:
```diff
@@ -0,0 +1,136 @@
+
+# Establish a connection to a DuckDB database located in the current directory. This connection allows for data modification as 'read_only' is set to FALSE.
+con <- dbConnect(duckdb::duckdb(), dbdir = "ARTIS_SAU_V0_03-08-2025_w_metadata.duckdb", read_only = FALSE)
+
```
This method of assigning data types to each column works well if we are going to use the data as dataframes in R. One of our goals by introducing duckdb to our workflow is to be efficient with local memory and standardize the duckdbtable schemas. Let's use the existing SQL table schemas to define the duckdb tables using regular SQL syntax. I wasn't clear on how this would work last week so I apologize for not communicating this more clearly before. 

Here is an example of what I am thinking:
```
library(DBI)
library(duckdb)

# 1. Create an empty table with a strict schema
dbExecute(con, "
    CREATE TABLE snet (
        record_id INTEGER PRIMARY KEY,
        importer_iso3c VARCHAR(3),
        exporter_iso3c VARCHAR(3),
        hs6 VARCHAR(6),
        product_weight_t DOUBLE,
        dom_source VARCHAR(100),
        source_country_iso3c VARCHAR(7),
        sciname VARCHAR(100),
        habitat VARCHAR(100),
        method VARCHAR(100),
        live_weight_t DOUBLE,
        hs_version VARCHAR(4),
        year INTEGER
    )
")

# 2. Import CSV data into the predefined table
dbExecute(con, "
    COPY snet FROM 'path/to/your_file.csv' 
    WITH (HEADER, DELIMITER ',', AUTO_DETECT FALSE)
")

## Can read parquet directly into duckdb table too
dbExecute(con, "
          CREATE TABLE tbl_artis AS
          SELECT * FROM read_parquet(
            '~/Documents/UW-SAFS/ARTIS/data/KNB_2024_07_31/data/artis_midpoint_all_HS_all_yrs_knb_v1.parquet')
          ")
```

## To-do

- [ ] Build tables in duckdb by passing `CREATE TABLE` and existing SQL schemas to duckdb via `DBI::dbExecute()` (I added links to SQL schemas in the table documentation) https://github.com/Seafood-Globalization-Lab/artis-model/wiki/ARTIS-Database-Tables#trade
- [ ] Import data for tables directly to duckdb by passing `COPY [...] FROM [...] WITH [...]` SQL via `DBI::dbExecute()` (see example above)

Anurag19101996 replied:
Hi @theamarks I will ask you more doubts on this in today's call.

Anurag19101996 replied:
Hi @theamarks 

1. I used CREATE TABLE to define blank SQL tables with strict column names and data types before loading any data.

2. Instead of using COPY FROM, I went with dbWriteTable() after coercing types in R so I could clean and validate the data first — it’s just a safer way to catch any mismatches or formatting issues before they hit the database. Its almost the same thing but do let me know if I should stick with COPY FROM.

The changes are made in `SAU_DuckDB_Script_w_metadata_v2.R`

theamarks commented on SAU Model/SAU_DuckDB_Script_w_metadata.R:
```diff
@@ -0,0 +1,136 @@
+trade_data$hs6 <- ifelse(nchar(trade_data$hs6) == 5, paste0("0", trade_data$hs6), trade_data$hs6)
+dbWriteTable(con, "trade", trade_data, overwrite = TRUE)
+
+
```
This is a great first step in inserting metadata directly into the duckdb, thank you. The next step is to make this more user friendly if each table had multiple rows. 

## To-do

- [ ] expand / make longer the `column_name`, `column_data_type`, and `column_description` columns so it is easy to scroll/search through the values. 
- [ ] Write metadata table with SQL schema (see example above)
```
CREATE TABLE metadata (
    table_name VARCHAR,  
    table_description
    column_name VARCHAR,  
    data_type VARCHAR, 
    column description VARCHAR 
);
```

Second

Anurag19101996 replied:
Hi @theamarks have a small doubt I will ask it in today's meeting.

Anurag19101996 replied:
Hi @theamarks 

I made the first changes in the script called `SAU_DuckDB_Script_w_metadata_v2.R'. I am a little confused for the second comment on what to do? Could you please elaborate?

theamarks reviewed (changes requested):
Thanks so much for your work @Anurag19101996. This is a great first step in building our new database for the ARTIS model. Thanks for your flexibility as we've been  brainstorming how this will work along the way. 

I've been thinking about how to align the different end-point resources we have for the ARTIS data (data repository archive, website, Github, duckdb, and the exploreARTIS package) to make it as easy and clear as possible for users. I have an idea that builds off of your work here and our conversations about the exploreARTIS package. The next step is to write a new exploreARTIS function that will download the ARTIS data tables directly from the KNB data repository and build a local .duckdb. This simplifies how we distribute the ARTIS .duckdb to users, and leverages our existing ARTIS infrastructure. I will outline this in a new issue and we can chat the next time we meet.

For now please finalize your `.R` here with the comments and checklists I left in the PR comments. Also feel free to message me in slack with any questions or tag me here in Github. 😎 🦆

Anurag19101996 committed:
File with creating sql tables to load data

# 6

Anurag19101996 reviewed (commented):
Hello @theamarks , could you review the changes for the new script I have uploaded.

Anurag19101996 committed:
Create process_knb_to_duckdb.R

R script to :
1. Download data from KNB
2. Unzip the files
3. Create the DuckDB

Anurag19101996 commented:
@theamarks added the script to get files from knb and convert it to a duckdb file

theamarks reviewed (approved):
Pull in Anurag's work to develop to document.
